# 확률 네트워크기반 친환경 자동차 충전소(GVCS) 로케이션 인텔리전스(Location Intelligence) 대시보드

## 제안 배경

**친환경 자동차 생산에만 관심이 쏠려 사용자를 고려하지 못한 설치로 인한 잉여 충전소 발생**

- 민관 협력을 통해 충전 서비스 관련 사업이 꾸준히 증가함. 하지만 전국에 설치된 공공 급속충전기 총 2,896개(2021년 기준) 중 하루 1회 이하로 사용한 충전기가 1,164개로 전체의 40.1%를 차지하고 있다. 심지어 17개는 전혀 사용하지 않은 것으로 나타남.

- 이와 같은 충전소마다 이용 편차가 큰 이유는 전기차 보급 목표에 따라 구역별로  충전소를 나눠 설치했기 때문이라는 지적도 있다.

- 한국전기차충전협동조합은 앞으로 수요가 몰리는 거점을 중심을 중심으로 충전기 설치를 해야하 한다는 의견 주장함.

- 기존의 정부 중심으로 공급자 입장에서 설치하던 충전소를 친환경 자동차 차주를 고려해 이용자 중심으로 변화 설치가 필요하다는 지적도 있다.

즉, **꾸준히 증가하는 친환경 차량의 수요를 위해 저비용 고효율의 충전소 설치를 목표 한다.** 충전소 설치 사업자의 기대 수익 최적화를 위해 사용자 입장을 고려한 체계적이며 최적의 입지 선정 결과를 제시함.

## 사용 데이터

1. 충전소 위치 및 현황 정보
2. 급속충전기 보급 현황
3. LPG충전소 현황
4. 오존 월별 대기오염도
5. 아황산가스 월별 대기오염도
6. 이산화질소 월별 대기오염도
7. 미세먼지 월별 대기오염도
8. 행정구역별 인구수
9. 항만 현황
10. 지역별 전기차 등록대 수

In [166]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import osmnx as ox
import networkx as nx
import numpy as np
from scipy.stats import norm
import plotly.express as px
import plotly.graph_objects as go

### 데이터 적합성 판단 기준

In [447]:
# 환경 요인 판단 기준
o3_standard = 0.003
so2_standard = 0.002
no2_standard = 0.012
co_standard = 0.3
pm10_standard = 15
pm25_standard = 10
khai_standard = 50

# 사회 요인 판단 기준
busan_people_standard = 200000
foottraffic_standard = 400000
elec_vehicle_standard = 50000
hidro_vehicle_standard = 1500
LPG_standard = 2
EVCS_standard = 500
HVCS_standard = 15
intersection_standard = 100

# 경제 요인 판단 기준
HVCS_construction_cost_standard = 3000000000
LPG_cost_standard = 6000000000
parking_area_standard = 100

# 기술 요인 판단 기준
# default 적합도 급속:0.38, 완속:0.62 

In [167]:
def advanced_replace(df, col, str, regex):
    '''
    DataFrame 특정 열의 특수문자 제거
    :param df: 대상 DataFrame
    :param col: column
    :param str: 제거할 특수 문자.
    :param regex: 적용할 정규표현식
    :return:
    '''
    df[col] = df[col].replace(str,0)
    df[col] = df[col].replace(to_replace=regex, value=r'', regex=True)
    return df

In [168]:
def show_norm(mean,std,min,max):
     '''
     정규분포 함수를 matplotlib을 통해 시각화
     mean: 평균값
     std: 표준편차
     min: 관측된 데이터의 최솟값
     max: 관측된 데이터의 최댓값
     '''
     min = min - std
     # x값 설정(min에서 max까지 unit만큼씩 증가하는 x)
     x = np.linspace(min,max,100)
     # y값은 평균이 mean이고 표준편차가 std인 확률밀도함수
     fig = px.line(x=x,y=norm.pdf(x,loc=mean,scale=std))
     fig.show()

In [169]:
def cal_norm(mean,std,min,max,value,affect):
     '''
     정규분포에서 value에 대한 누적확률 구하기
     mean: 평균값
     std: 표준편차
     min: 관측된 데이터의 최솟값
     max: 관측된 데이터의 최댓값
     value: 누적확률을 구하고자 하는 데이터의 측정값
     affect: 입지에 긍정/부정적인 영향(True/False)
     '''
     min = min - std
     red ='rgb(239,85,59)'
     blue = 'rgb(100,110,250)'
     fig = go.Figure()
     
     # 확률 값을 구할 특정 구간의 범위 설정
     cum_a = np.linspace(min,value,100)
     cum_b = np.linspace(value,max,100)
     
     pro = norm(mean,std).cdf(value)-norm(mean,std).cdf(min).round(3)
     if affect == True:
          # 구간 사이에 색을 채움
          fig.add_trace(go.Scatter(x=cum_a,y=norm.pdf(cum_a,mean,std), fill = 'tozeroy',name='적합',line=dict(color=blue)))
          fig.add_trace(go.Scatter(x=cum_b,y=norm.pdf(cum_b,mean,std), fill = 'tozeroy',name='부적합',line=dict(color=red)))
          fig.update_yaxes(visible=False)
          annotations = []
          annotations.append(dict(x=value, y=norm.pdf(value,loc=mean,scale=std), showarrow=False, text=round(pro,3), font=dict(size=15,color=blue), xshift=-40, yshift=-100, bordercolor=blue,borderwidth=2))
          annotations.append(dict(x=value, y=norm.pdf(value,loc=mean,scale=std), showarrow=False, text=round(1-pro,3),font=dict(size=15,color=red),xshift=40, yshift=-100, bordercolor=red,borderwidth=2))
     else:
          fig.add_trace(go.Scatter(x=cum_a,y=norm.pdf(cum_a,mean,std), fill = 'tozeroy',name='부적합',line=dict(color=red)))
          fig.add_trace(go.Scatter(x=cum_b,y=norm.pdf(cum_b,mean,std), fill = 'tozeroy',name='적합',line=dict(color=blue)))
          fig.update_yaxes(visible=False)
          annotations = []
          annotations.append(dict(x=value, y=norm.pdf(value,loc=mean,scale=std), showarrow=False, text=round(pro,3), font=dict(size=15,color=red), xshift=-40, yshift=-100, bordercolor=red,borderwidth=2))
          annotations.append(dict(x=value, y=norm.pdf(value,loc=mean,scale=std), showarrow=False, text=round(1-pro,3),font=dict(size=15,color=blue),xshift=40, yshift=-100, bordercolor=blue,borderwidth=2))     
          pro = 1 - pro
     fig.update_layout(annotations=annotations)
     # value까지의 누적확률에서 min까지의 누적확률을 뺌
     fig.show()

     # 최종 누적확률 반환
     return pro, 1-pro

## 상위 요소 - 환경적 요인

대기오염도를 기반하여 확률 값을 계산한다.

### 대기오염도 데이터 불러오기

In [226]:
air_pollution_file_path = "C:\\Users\\SAMSUNG\\Desktop\\Green-Car-Charging-Station-Site-Selection\\Data\\대기오염시도별실시간측정정보(환경부).csv"
df_air_pollution = pd.read_csv(air_pollution_file_path, encoding='cp949')
df_air_pollution

sidoName stationName mangName  so2Value  coValue  o3Value  no2Value  \
0           충북    송정동(봉명동)     도시대기     0.002      0.6    0.003     0.022   
1           충북         사천동     도시대기     0.002      0.5    0.002     0.019   
2           충북         용담동     도시대기       NaN      0.4    0.002     0.018   
3           충북         용암동     도시대기     0.002      0.5    0.002     0.017   
4           충북         복대동    도로변대기     0.004      0.7    0.002     0.028   
...        ...         ...      ...       ...      ...      ...       ...   
10907       강원         묵호항       항만     0.013      1.3    0.003     0.034   
10908       강원     철원(DMZ)     교외대기     0.001      0.3    0.023     0.010   
10909       강원     화천(DMZ)     교외대기     0.001      0.3    0.033     0.007   
10910       강원     인제(DMZ)     교외대기     0.001      0.2    0.038     0.005   
10911       강원     고성(DMZ)     교외대기     0.002      0.2    0.044     0.005   

       pm10Value  pm25Value  khaiValue  
0           65.0       12.0       56.0  
1           28.0       22.0       59.0  
2           12.0       11.0       30.0  
3           31.0       15.0       52.0  
4           25.0       10.0       47.0  
...          ...        ...        ...  
10907       33.0       16.0       56.0  
10908       28.0       27.0       82.0  
10909       16.0       15.0       64.0  
10910       10.0        8.0       57.0  
10911       20.0       12.0       62.0  

[10912 rows x 10 columns]

### 결측값 보간

In [227]:
df_air_pollution = df_air_pollution.interpolate(method='pad')
df_air_pollution

sidoName stationName mangName  so2Value  coValue  o3Value  no2Value  \
0           충북    송정동(봉명동)     도시대기     0.002      0.6    0.003     0.022   
1           충북         사천동     도시대기     0.002      0.5    0.002     0.019   
2           충북         용담동     도시대기     0.002      0.4    0.002     0.018   
3           충북         용암동     도시대기     0.002      0.5    0.002     0.017   
4           충북         복대동    도로변대기     0.004      0.7    0.002     0.028   
...        ...         ...      ...       ...      ...      ...       ...   
10907       강원         묵호항       항만     0.013      1.3    0.003     0.034   
10908       강원     철원(DMZ)     교외대기     0.001      0.3    0.023     0.010   
10909       강원     화천(DMZ)     교외대기     0.001      0.3    0.033     0.007   
10910       강원     인제(DMZ)     교외대기     0.001      0.2    0.038     0.005   
10911       강원     고성(DMZ)     교외대기     0.002      0.2    0.044     0.005   

       pm10Value  pm25Value  khaiValue  
0           65.0       12.0       56.0  
1           28.0       22.0       59.0  
2           12.0       11.0       30.0  
3           31.0       15.0       52.0  
4           25.0       10.0       47.0  
...          ...        ...        ...  
10907       33.0       16.0       56.0  
10908       28.0       27.0       82.0  
10909       16.0       15.0       64.0  
10910       10.0        8.0       57.0  
10911       20.0       12.0       62.0  

[10912 rows x 10 columns]

### 부산 대기오염 데이터 추출

In [231]:
df_air_pollution = df_air_pollution[df_air_pollution['sidoName']=='부산']
df_air_pollution.reset_index(drop=True,inplace=True)
df_air_pollution

sidoName stationName mangName  so2Value  coValue  o3Value  no2Value  \
0         부산         광복동     도시대기     0.002      0.3    0.016     0.018   
1         부산         초량동    도로변대기     0.002      0.4    0.010     0.028   
2         부산         태종대     도시대기     0.003      0.2    0.019     0.020   
3         부산         청학동     도시대기     0.003      0.2    0.021     0.018   
4         부산         전포동     도시대기     0.002      0.2    0.016     0.014   
..       ...         ...      ...       ...      ...      ...       ...   
625       부산         하단동     도시대기     0.003      0.4    0.015     0.029   
626       부산         하단동     도시대기     0.003      0.3    0.023     0.015   
627       부산         하단동     도시대기     0.002      0.5    0.005     0.031   
628       부산         하단동     도시대기     0.004      0.5    0.002     0.047   
629       부산         하단동     도시대기     0.004      0.8    0.002     0.061   

     pm10Value  pm25Value  khaiValue  
0         11.0        8.0       30.0  
1         16.0        5.0       47.0  
2         15.0       10.0       33.0  
3         10.0        7.0       35.0  
4         14.0       10.0       37.0  
..         ...        ...        ...  
625       43.0       23.0       66.0  
626       31.0       22.0       59.0  
627       25.0       17.0       51.0  
628       27.0       17.0       78.0  
629       32.0       26.0      101.0  

[630 rows x 10 columns]

#### 부산시 오존 적합도

In [237]:
o3_pro,o3_pro_negative = cal_norm(df_air_pollution['o3Value'].mean(),df_air_pollution['o3Value'].std(),
df_air_pollution['o3Value'].min(),df_air_pollution['o3Value'].max(),o3_standard,False)
o3_pro

0.8017202428389887

#### 하단동의 오존대기오염 확률 계산

In [238]:
ozone_pro,ozone_pro_negative = cal_norm(df_air_pollution['o3Value'].mean(),df_air_pollution['o3Value'].std(),
df_air_pollution['o3Value'].min(),df_air_pollution['o3Value'].max(),df_air_pollution[df_air_pollution['stationName']=='하단동'].tail(1)['o3Value'].values[0],True)
ozone_pro

0.17217590368470223

#### 부산시 오존 적합도

In [242]:
so2_pro,so2_pro_negative = cal_norm(df_air_pollution['so2Value'].mean(),df_air_pollution['so2Value'].std(),
df_air_pollution['so2Value'].min(),df_air_pollution['so2Value'].max(),so2_standard,False)
so2_pro

0.7911203783441816

#### 부산시 이산화질소 적합도

In [243]:
no2_pro,no2_pro_negative = cal_norm(df_air_pollution['no2Value'].mean(),df_air_pollution['no2Value'].std(),
df_air_pollution['no2Value'].min(),df_air_pollution['no2Value'].max(),no2_standard,False)
no2_pro

0.909590433996196

#### 부산시 일산화탄소 적합도

In [376]:
co_pro,co_pro_negative = cal_norm(df_air_pollution['coValue'].mean(),df_air_pollution['coValue'].std(),
df_air_pollution['coValue'].min(),df_air_pollution['coValue'].max(),co_standard,False)
co_pro

0.8287531282163694

#### 부산시 미세먼지 적합도

In [246]:
pm10_pro,pm10_pro_negative = cal_norm(df_air_pollution['pm10Value'].mean(),df_air_pollution['pm10Value'].std(),
df_air_pollution['pm10Value'].min(),df_air_pollution['pm10Value'].max(),pm10_standard,False)
pm10_pro

0.8866811666335976

In [247]:
pm25_pro,pm25_pro_negative = cal_norm(df_air_pollution['pm25Value'].mean(),df_air_pollution['pm25Value'].std(),
df_air_pollution['pm25Value'].min(),df_air_pollution['pm25Value'].max(),pm25_standard,False)
pm25_pro

0.8415688322677302

#### 부산시 통합대기환경 지수 적합도

In [278]:
khai_pro,khai_pro_negative = cal_norm(df_air_pollution['khaiValue'].mean(),df_air_pollution['khaiValue'].std(),
df_air_pollution['khaiValue'].min(),df_air_pollution['khaiValue'].max(),khai_standard,False)
khai_pro

0.8220286272881268

## 상위 요소 - 사회적 요인

### 행정구역 인구 데이터 불러오기

In [180]:
people_file_path = 'C:\\Users\\SAMSUNG\\Desktop\\Green-Car-Charging-Station-Site-Selection\\Data\\행정구역_시군구_별__성별_인구수.csv'
df_people = pd.read_csv(people_file_path, encoding='cp949', header=1)
df_people

행정구역(시군구)별  총인구수 (명)
0           전국  51592660
1        서울특별시   9500480
2          종로구    144005
3           중구    122087
4          용산구    222009
..         ...       ...
287        거창군     60756
288        합천군     42613
289    제주특별자치도    677413
290        제주시    493076
291       서귀포시    184337

[292 rows x 2 columns]

### 부산광역시 인구 데이터 추출

In [181]:
df_busan_people = df_people[28:44].sort_values('행정구역(시군구)별')
df_busan_people.set_index('행정구역(시군구)별',inplace=True)
df_busan_people['총인구수 (명)'] = df_busan_people['총인구수 (명)'].astype(float)
df_busan_people

총인구수 (명)
행정구역(시군구)별          
강서구         143638.0
금정구         225033.0
기장군         176505.0
남구          260617.0
동구           88257.0
동래구         273976.0
부산진구        351022.0
북구          280852.0
사상구         206365.0
사하구         305556.0
서구          103961.0
수영구         174633.0
연제구         206677.0
영도구         109407.0
중구           40585.0
해운대구        393130.0

In [182]:
df_busan_people.describe()

총인구수 (명)
count      16.000000
mean   208763.375000
std     98697.592227
min     40585.000000
25%    135080.250000
50%    206521.000000
75%    275695.000000
max    393130.000000

#### 부산시 고정인구 적합도

In [254]:
busan_people_pro,busan_people_pro_negative = cal_norm(df_busan_people['총인구수 (명)'].mean(), df_busan_people['총인구수 (명)'].std(), df_busan_people['총인구수 (명)'].min(), 
df_busan_people['총인구수 (명)'].max(), busan_people_standard, False)

### 유동인구 데이터 불러오기

In [184]:
foottraffic_file_path = 'C:\\Users\\SAMSUNG\\Desktop\\Green-Car-Charging-Station-Site-Selection\\Data\\부산 유동인구 데이터.csv'
df_foottraffic = pd.read_csv(foottraffic_file_path, encoding='cp949')
df_foottraffic

구군          1월          2월          3월          4월          5월  \
0    강서구  325,284.39  323,049.87  335,120.06  340,570.52  332,289.77   
1    금정구  482,598.89  477,047.11  484,669.59  490,864.46  487,435.45   
2    기장군  341,181.17  343,398.79  342,489.52  345,184.20  348,793.68   
3     남구  539,072.58  526,693.91  524,611.67  524,961.53  526,758.54   
4     동구  185,466.33  185,459.30  189,082.61  192,702.25  189,819.14   
5    동래구  553,655.76  549,718.60  543,091.96  535,439.53  533,908.61   
6   부산진구  757,261.10  760,742.04  760,212.63  761,728.33  765,460.09   
7     북구  559,133.62  554,946.17  543,084.57  543,479.25  543,145.91   
8    사상구  496,102.34  492,871.78  493,054.71  496,932.63  494,094.59   
9    사하구  641,541.35  632,045.91  628,461.73  633,805.98  630,333.83   
10    서구  244,816.18  247,144.65  246,832.32  249,028.55  245,968.42   
11   수영구  339,989.93  335,887.41  345,588.95  358,682.04  356,153.28   
12   연제구  394,132.20  390,539.12  393,604.79  396,276.36  394,296.80   
13   영도구  248,178.99  246,981.20  242,904.23  243,650.92  245,374.40   
14    중구  105,170.15  106,603.51  110,729.44  110,668.93  113,610.04   
15  해운대구  867,047.22  871,849.92  858,675.53  862,845.41  867,445.21   

            6월          7월          8월          9월         10월         11월  \
0   340,678.45  337,488.64  328,374.74  328,764.38  333,414.26  340,405.10   
1   484,936.00  483,898.58  482,964.23  487,082.06  484,633.90  485,502.59   
2   348,500.32  350,763.40  349,328.59  348,077.97  348,449.77  345,905.89   
3   531,078.89  529,810.42  530,199.16  531,658.35  525,870.37  528,106.40   
4   191,618.73  189,864.16  189,368.74  190,328.67  191,182.03  194,732.77   
5   528,124.07  527,730.64  525,468.79  532,656.70  527,498.20  520,126.56   
6   764,636.01  759,638.81  754,837.78  763,942.81  761,731.75  763,818.34   
7   540,381.96  542,289.98  546,905.11  551,394.74  542,887.88  539,732.97   
8   494,207.66  492,973.61  490,296.85  493,842.47  491,808.21  493,815.86   
9   629,635.14  630,260.69  631,377.03  633,402.47  627,551.97  625,380.25   
10  232,053.80  230,458.90  229,377.92  228,929.31  225,006.09  222,113.90   
11  344,222.04  345,425.50  342,987.83  342,228.84  340,723.32  340,813.55   
12  399,403.98  401,469.81  400,414.71  398,036.79  395,250.20  398,436.74   
13  244,196.20  243,830.89  244,457.54  244,665.84  242,981.94  241,042.77   
14  128,260.41  127,285.37  124,056.84  127,550.96  131,073.84  135,646.63   
15  867,936.68  875,772.89  871,029.74  864,691.10  864,549.71  863,913.89   

           12월        월 평균  
0   337,753.15  333,599.44  
1   488,392.99  485,002.15  
2   344,245.29  346,359.88  
3   531,610.18  529,202.67  
4   193,341.86  190,247.22  
5   524,788.88  533,517.36  
6   768,957.14  761,913.90  
7   545,952.52  546,111.22  
8   495,297.44  493,774.85  
9   631,475.21  631,272.63  
10  223,835.54  235,463.80  
11  345,960.31  344,888.58  
12  401,486.80  396,945.69  
13  240,790.94  244,087.99  
14  134,834.02  121,290.85  
15  878,342.31  867,841.63

In [185]:
foottraffic_col = df_foottraffic.iloc[:,1:-1].columns.tolist()

In [186]:
df_foottraffic.drop(columns=foottraffic_col, inplace=True)
df_foottraffic.set_index('구군',inplace=True)
df_foottraffic

월 평균
구군              
강서구   333,599.44
금정구   485,002.15
기장군   346,359.88
남구    529,202.67
동구    190,247.22
동래구   533,517.36
부산진구  761,913.90
북구    546,111.22
사상구   493,774.85
사하구   631,272.63
서구    235,463.80
수영구   344,888.58
연제구   396,945.69
영도구   244,087.99
중구    121,290.85
해운대구  867,841.63

In [187]:
advanced_replace(df_foottraffic, '월 평균','-',r'[^0-9.0-9]')
df_foottraffic['월 평균'] = df_foottraffic['월 평균'].astype(float)

In [188]:
df_foottraffic.describe()

월 평균
count      16.000000
mean   441344.991250
std    205145.538509
min    121290.850000
25%    311221.577500
50%    440973.920000
75%    536665.825000
max    867841.630000

In [253]:
foottraffic_pro,foottraffic_pro_negative = cal_norm(df_foottraffic['월 평균'].mean(),df_foottraffic['월 평균'].std(),
df_foottraffic['월 평균'].min(),df_foottraffic['월 평균'].max(),foottraffic_standard,False)
foottraffic_pro


0.5848617448890694

### 친환경 자동차 등록수 데이터 불러오기

In [191]:
vehicle_file_path = 'C:\\Users\\SAMSUNG\\Desktop\\Green-Car-Charging-Station-Site-Selection\\Data\\연료별 자동차 등록 수.csv'
df_vehicle = pd.read_csv(vehicle_file_path, encoding='cp949')
drop_col = df_vehicle.iloc[:,1:3].columns.tolist()
df_vehicle.drop(columns=drop_col,inplace=True)
df_vehicle

연료         서울         부산         대구         인천       광주  \
0            휘발유  1,649,518    714,221    584,780    818,874  312,081   
1             경유  1,077,183    550,411    449,206    627,029  266,237   
2            엘피지    240,940    111,371    106,654    113,654   88,883   
3             전기     45,133     16,059     18,076     15,191    6,233   
4  하이브리드(휘발유+전기)    147,022     61,167     48,320     90,275   24,091   
5   하이브리드(경유+전기)      2,656      4,187      2,146      7,672      464   
6  하이브리드(LPG+전기)      1,601        735        806        696      628   
7           기타연료      7,172     16,097      4,545     13,915    4,036   
8             수소      2,435      1,473        352      1,139    1,018   
9             총합  3,173,660  1,475,721  1,214,885  1,688,445  703,671   

        대전       울산       세종         경기       강원       충북         충남       전북  \
0  341,412  297,231   94,929  3,056,798  365,724  403,958    525,588  398,182   
1  247,724  217,553   66,126  2,450,214  371,277  380,402    516,001  430,676   
2   67,456   36,101   12,985    404,449   58,335   70,304     89,210   93,615   
3    9,550    3,948    2,180     50,111    9,117   10,421     12,142    8,931   
4   24,167   19,593   11,034    256,643   24,001   25,842     32,669   25,448   
5      393      194      141      3,964      286      580        461      694   
6      481      457      153      3,453      611      639        728      785   
7    3,830    5,205    1,208     29,809    4,376    6,918      9,574    5,182   
8    1,045    2,395      205      4,004    1,959    1,095        908    1,435   
9  696,058  582,677  188,961  6,259,445  835,686  900,159  1,187,281  964,948   

          전남         경북         경남       제주           계  
0    466,016    654,776    886,361  302,434  11,872,883  
1    547,999    664,042    736,135  253,205   9,851,420  
2    111,494    127,818    139,592   58,412   1,931,273  
3     10,658     13,018     14,957   26,931     272,656  
4     38,154     36,999     56,219   30,057     951,701  
5        486        405      1,774      341      26,844  
6        833        974      1,034      335      14,949  
7      9,666     10,974      9,161      886     142,554  
8        448         77      1,899        5      21,892  
9  1,185,754  1,509,083  1,847,132  672,606  25,086,172

In [192]:
# 전국 자동차 수 통계값 구하기 (전기, 수소)
col = df_vehicle.iloc[:,1:].columns.tolist()
for c in col:
    advanced_replace(df_vehicle,c,'-',r'[^0-9.0-9]')
    df_vehicle[c] = df_vehicle[c].astype(int)
df_elec_vehicle = df_vehicle.iloc[3:7,:-1].sum()
df_elec_vehicle = df_elec_vehicle[1:]

df_elec_vehicle = df_elec_vehicle.astype(int)
df_elec_vehicle.describe()


count        17.000000
mean      74479.411765
std       75151.087588
min       13508.000000
25%       34591.000000
50%       50131.000000
75%       73984.000000
max      314171.000000
dtype: float64

In [193]:
df_hidro_vehicle = df_vehicle.iloc[8,1:-1]
df_hidro_vehicle = df_hidro_vehicle.astype(int)
df_hidro_vehicle.describe()

count      17.000000
mean     1287.764706
std      1030.779531
min         5.000000
25%       448.000000
50%      1095.000000
75%      1899.000000
max      4004.000000
Name: 8, dtype: float64

#### 부산시 전기차, 수소차 수 적합도

In [252]:
elec_vehicle_pro,elec_vehicle_pro_negative = cal_norm(df_elec_vehicle.mean(), df_elec_vehicle.std(),
 df_elec_vehicle.min(), df_elec_vehicle.max(),elec_vehicle_standard,False)
print(elec_vehicle_pro)

hidro_vehicle_pro,hidro_vehicle_pro_negative = cal_norm(df_hidro_vehicle.mean(), df_hidro_vehicle.std(),
 df_hidro_vehicle.min(), df_hidro_vehicle.max(),hidro_vehicle_standard,False)
print(hidro_vehicle_pro)

0.6626879332080854


0.4304353506049742


### LPG 충전소 현황 데이터 불러오기 (수소차)

In [196]:
LPG_file_path = 'C:\\Users\\SAMSUNG\\Desktop\\Green-Car-Charging-Station-Site-Selection\\Data\\부산 LPG 충전소 현황(한국가스안전공사).csv'
df_LPG = pd.read_csv(LPG_file_path, encoding='cp949')
df_LPG

행정구역                   업소명                  소재지              관리구분  \
0    부산 강서구            개인택시신공항충전소  부산 강서구 경전철로188번길 95             자동차충전   
1    부산 강서구            부산공항LPG충전소       부산 강서구 공항로 459             자동차충전   
2    부산 강서구                신항만에너지     부산 강서구 낙동남로 507              자동차충전   
3    부산 강서구  (주)지원네트웍스 부산강서LPG충전소     부산 강서구 낙동북로 310              자동차충전   
4    부산 강서구                신명지에너지    부산 강서구 르노삼성대로 560             자동차충전   
..      ...                   ...                  ...               ...   
58   부산 연제구          부산개인택시엘피지충전소     부산 연제구 월드컵대로 399             자동차충전   
59   부산 영도구              (주)태종에너지       부산 영도구 영선대로 41             자동차충전   
60   부산 영도구                 부산에너지       부산 영도구 해양로 160  용기+자동차+탱크+13kg용기   
61   부산 영도구         개인택시 영도LPG충전소       부산 영도구 해양로 296             자동차충전   
62  부산 해운대구          이원 해운대LPG충전소    부산 해운대구 해운대로 1255  용기+자동차+탱크+13kg용기   

           위도          경도      면적       공시지가  
0   35.198669  128.966594  2545.0  2144000.0  
1   35.149064  128.951891  3105.0  1174000.0  
2   35.112947  128.875260  3105.0  1174000.0  
3   35.212507  128.962572  1396.0  1140000.0  
4   35.097554  128.927351   481.0   700300.0  
..        ...         ...     ...        ...  
58  35.186013  129.055974  9212.0   688600.0  
59  35.084996  129.040033  1610.2  2679000.0  
60  35.091275  129.069484  5801.0   989400.0  
61  35.082176  129.076429  1818.0   349800.0  
62  35.196728  129.205260  3700.0  1607000.0  

[63 rows x 8 columns]

In [197]:
# 복합충전소를 위해 1500 m^2이상의 부지가 요구됨
df_LPG = df_LPG[df_LPG['면적'] >= 1500].reset_index(drop=True)
df_LPG

행정구역                      업소명                      소재지  \
0    부산 강서구               개인택시신공항충전소      부산 강서구 경전철로188번길 95   
1    부산 강서구               부산공항LPG충전소           부산 강서구 공항로 459   
2    부산 강서구                   신항만에너지         부산 강서구 낙동남로 507    
3    부산 강서구               르노삼성자동차(주)         부산 강서구 르노삼성대로 61   
4    부산 강서구                 죽림LPG충전소          부산 강서구 호계로 102    
5    부산 강서구                    가락충전소          부산 강서구 호계로 195    
6    부산 금정구                    (주)해연       부산 금정구 개좌로 225-15    
7    부산 금정구                  (주)신진가스          부산 금정구 반송로 433    
8    부산 기장군            마라톤가스(주)청강충전소   부산 기장군 기장읍 기장대로 343 A동   
9    부산 기장군               한국에너지산업(주)  부산 기장군 기장읍 기장대로413번길 38   
10   부산 기장군                  선암가스충전소    부산 기장군 장안읍 기장대로 1451    
11   부산 기장군    KH에너지(주)직영장안충전소(부산방향)     부산 기장군 장안읍 동해고속도로 25   
12   부산 기장군    KH에너지(주)직영장안충전소(울산방향)     부산 기장군 장안읍 동해고속도로 26   
13   부산 기장군                    정림에너지     부산 기장군 정관읍 신정관로 187    
14   부산 기장군        광신개발(주)직영 정관가스충전소       부산 기장군 정관읍 정관로 907   
15   부산 기장군               개인택시동부산충전소      부산 기장군 철마면 반송로 1022   
16    부산 동구             개인택시남부LPG충전소             부산 동구 성남로 42   
17    부산 동구  (주)대양씨앤씨(구:녹색에너지 범일영업소)           부산 동구 성남로 74-3   
18    부산 북구              개인택시 금곡역충전소           부산 북구 금곡대로 704   
19    부산 북구                    (주)삼보         부산 북구 만덕대로 327-0   
20   부산 사상구                  팔성가스충전소           부산 사상구 가야대로 75   
21   부산 사상구            (주)성은에너지제2충전소          부산 사상구 강변대로 412   
22   부산 사상구             극동유화(주)감전충전소          부산 사상구 낙동대로 878   
23   부산 사하구             광진기업(주)장림영업소           부산 사하구 다대로 309   
24   부산 사하구          광신가스(주)을숙도대로충전소         부산 사하구 을숙도대로 530   
25   부산 사하구             개인택시사하엘피지충전소         부산 사하구 을숙도대로 583   
26   부산 연제구             부산개인택시엘피지충전소         부산 연제구 월드컵대로 399   
27   부산 영도구                 (주)태종에너지           부산 영도구 영선대로 41   
28   부산 영도구                    부산에너지           부산 영도구 해양로 160   
29   부산 영도구            개인택시 영도LPG충전소           부산 영도구 해양로 296   
30  부산 해운대구             이원 해운대LPG충전소        부산 해운대구 해운대로 1255   

                관리구분         위도          경도       면적       공시지가  
0              자동차충전  35.198669  128.966594   2545.0  2144000.0  
1              자동차충전  35.149064  128.951891   3105.0  1174000.0  
2              자동차충전  35.112947  128.875260   3105.0  1174000.0  
3              자동차충전  35.095408  128.883215   1538.0   640100.0  
4              자동차충전  35.200736  128.893462   1580.0  1017000.0  
5              자동차충전  35.208847  128.890148   2600.0  1017000.0  
6       용기+탱크+13kg용기  35.230796  129.122388   3346.0  1370000.0  
7              자동차충전  35.215006  129.111143   1646.4  2797000.0  
8              자동차충전  35.225657  129.213988   2592.0  1069000.0  
9   용기+자동차+탱크+13kg용기  35.235543  129.215629   2473.0  1063000.0  
10        자동차+13kg용기  35.321585  129.241320   2717.0  1060000.0  
11             자동차충전  35.377525  129.246402  15637.0   263500.0  
12             자동차충전  35.382870  129.248394  37294.0   263500.0  
13             자동차충전  35.328714  129.201710   2575.0   743500.0  
14        자동차+13kg용기  35.324749  129.210964   1899.0   877400.0  
15             자동차충전  35.235930  129.163728   2543.0   923000.0  
16             자동차충전  35.130627  129.054917   1518.5  1701000.0  
17         용기+13kg용기  35.130891  129.057587   3169.8  1545000.0  
18             자동차충전  35.269475  129.018781   2055.0   842500.0  
19             자동차충전  35.213643  129.040635   3837.0  1608000.0  
20        자동차+13kg용기  35.151317  128.980986   2946.6  3247000.0  
21             자동차충전  35.122092  128.962063   3732.4  2373000.0  
22  용기+자동차+탱크+13kg용기  35.138888  128.971339   2886.1  3201000.0  
23     용기+자동차+13kg용기  35.075007  128.977163   1604.0  2665000.0  
24             자동차충전  35.085225  128.963646   1632.0  1095000.0  
25             자동차충전  35.085534  128.969349   1657.0  1498000.0  
26             자동차충전  35.186013  129.055974   9212.0   688600.0  
27             자동차충전  35.084996  129.040033   1610.2  2679000.0  
28  용기+자동차+탱

In [198]:
LPG_group = df_LPG.groupby('행정구역').count()
LPG_group.describe()

업소명        소재지       관리구분         위도         경도         면적  \
count  10.000000  10.000000  10.000000  10.000000  10.000000  10.000000   
mean    3.100000   3.100000   3.100000   3.100000   3.100000   3.100000   
std     2.233582   2.233582   2.233582   2.233582   2.233582   2.233582   
min     1.000000   1.000000   1.000000   1.000000   1.000000   1.000000   
25%     2.000000   2.000000   2.000000   2.000000   2.000000   2.000000   
50%     2.500000   2.500000   2.500000   2.500000   2.500000   2.500000   
75%     3.000000   3.000000   3.000000   3.000000   3.000000   3.000000   
max     8.000000   8.000000   8.000000   8.000000   8.000000   8.000000   

            공시지가  
count  10.000000  
mean    3.100000  
std     2.233582  
min     1.000000  
25%     2.000000  
50%     2.500000  
75%     3.000000  
max     8.000000

#### LPG 충전소 수 적합도

In [251]:
LPG_pro,LPG_pro_negative = cal_norm(LPG_group['업소명'].mean(),LPG_group['업소명'].std(),
LPG_group['업소명'].min(),LPG_group['업소명'].max(),LPG_standard,False)

### 전기차 충전소 현황 데이터 불러오기

In [200]:
EVCS_file_path = 'C:\\Users\\SAMSUNG\\Desktop\\Green-Car-Charging-Station-Site-Selection\\Data\\부산 전기차 충전소 정보.csv'
df_EVCS = pd.read_csv(EVCS_file_path, encoding='cp949')
df_EVCS

운영기관         충전소  충전기ID            충전기타입      지역   시군구  \
0     환경부(한국자동차환경협회)  kbs 부산방송총국      1  DC차데모+AC3상+DC콤보  부산광역시   수영구    
1     환경부(한국자동차환경협회)  kbs 부산방송총국     11             DC콤보  부산광역시   수영구    
2     환경부(한국자동차환경협회)  kbs 부산방송총국     12             DC콤보  부산광역시   수영구    
3     환경부(한국자동차환경협회)     국립부산과학관     11             DC콤보  부산광역시   기장군    
4     환경부(한국자동차환경협회)     국립부산과학관     12             DC콤보  부산광역시   기장군    
...              ...         ...    ...              ...     ...   ...   
6291          한국전자금융    부산 플러스타워      1             AC완속  부산광역시   강서구    
6292          한국전자금융    부산 플러스타워      2             AC완속  부산광역시   강서구    
6293          한국전자금융   부산 휴마트괴정점      1             AC완속  부산광역시   사하구    
6294          한국전자금융   부산 휴마트괴정점      2             AC완속  부산광역시   사하구    
6295          한국전자금융   부산 휴마트괴정점      3             AC완속  부산광역시   사하구    

                                            주소         위도          경도  
0                            부산광역시 수영구 수영로 429  35.144806  129.107595  
1                            부산광역시 수영구 수영로 429  35.144806  129.107595  
2                            부산광역시 수영구 수영로 429  35.144806  129.107595  
3                        부산광역시 기장군 기장읍 석산리 293  35.204002  129.212987  
4                        부산광역시 기장군 기장읍 석산리 293  35.204002  129.212987  
...                                        ...        ...         ...  
6291  부산광역시 강서구 명지국제1로 56-1플러스타워(3598-2 플러스타워)  35.093685  128.905372  
6292  부산광역시 강서구 명지국제1로 56-1플러스타워(3598-2 플러스타워)  35.093685  128.905372  
6293       부산광역시 사하구 낙동대로 123휴마트괴정점(괴정동 358-1)  35.103200  129.002528  
6294       부산광역시 사하구 낙동대로 123휴마트괴정점(괴정동 358-1)  35.103200  129.002528  
6295       부산광역시 사하구 낙동대로 123휴마트괴정점(괴정동 358-1)  35.103200  129.002528  

[6296 rows x 9 columns]

In [202]:
EVCS_group = df_EVCS.groupby('시군구').count()
EVCS_group

운영기관   충전소  충전기ID  충전기타입    지역    주소    위도    경도
시군구                                                    
강서구     465   465    465    465   465   465   465   465
금정구     311   311    311    311   311   311   311   311
기장군     504   504    504    504   504   504   504   504
남구      681   681    681    681   681   681   681   681
동구      119   119    119    119   119   119   119   119
동래구     471   471    471    471   471   471   471   471
부산진구    583   583    583    583   583   583   583   583
북구      353   353    353    353   353   353   353   353
사상구     315   315    315    315   315   315   315   315
사하구     346   346    346    346   346   346   346   346
서구      133   133    133    133   133   133   133   133
수영구     253   253    253    253   253   253   253   253
연제구     342   342    342    342   342   342   342   342
영도구     150   150    150    150   150   150   150   150
중구       28    28     28     28    28    28    28    28
해운대구   1237  1237   1237   1237  1237  1237  1237  1237

#### 전기차 충전소 수 적합도

In [250]:
EVCS_group['충전소'] = EVCS_group['충전소'].astype(int)
EVCS_pro,EVCS_pro_negative = cal_norm(EVCS_group['충전소'].mean(),EVCS_group['충전소'].std(),
EVCS_group['충전소'].min(),EVCS_group['충전소'].max(),EVCS_standard,True)

### 수소충전소 현황 데이터 불러오기

In [204]:
HVCS_file_path = 'C:\\Users\\SAMSUNG\\Desktop\\Green-Car-Charging-Station-Site-Selection\\Data\\수소충전소(환경부).csv'
df_HVCS = pd.read_csv(HVCS_file_path, encoding='cp949')
df_HVCS

충전소  충전기                       주소  \
0          상암수소충전소    1     서울특별시 마포구 하늘공원로 108    
1        마곡에코수소충전소    1       서울특별시 강서구 양천로 225    
2          오곡수소충전소    1       서울특별시 강서구 벌말로 271    
3         H국회수소충전소    1      서울특별시 영등포구 의사당대로 1    
4          양재수소충전소    1    서울특별시 서초구 바우뫼로12길 73    
..             ...  ...                      ...   
103  하이넷 통영용남수소충전소    1   경상남도 통영시 용남면 남해안대로 38    
104        김해수소충전소    2  경상남도 김해시 김해대로2596번길 53    
105  하이넷 거제장평수소충전소    1      경상남도 거제시 거제대로 5050    
106     양산시증산수소충전소    1    경상남도 양산시 물금읍 메기로 114    
107       H함안수소충전소    1    경상남도 함안군 군북면 현포로 205    

                            이용가능시간  
0                      8~20시(연중무휴)  
1                    8~20시(일요일 휴무)  
2                      8~22시(연중무휴)  
3         평일 8~24시 / 주말, 공휴일 8~22시  
4                      8~22시(연중무휴)  
..                             ...  
103  7~21시(매주 수요일 / 둘, 넷째주 일요일 휴무)  
104       평일 8~21시 / 주말, 공휴일 9~18시  
105                  6~24시(일요일 휴무)  
106                    7~23시(연중무휴)  
107                    8~22시(연중무휴)  

[108 rows x 4 columns]

In [205]:
address_city = []
address_gu = []
for i in range(df_HVCS.shape[0]):
    strings = df_HVCS.iloc[i]['주소'].split()
    address_city.append(strings.pop(0))
    address_gu.append(strings.pop(0))
df_HVCS.insert(2,"시",address_city)
df_HVCS.insert(3,"구",address_gu)    
df_HVCS

충전소  충전기      시     구                       주소  \
0          상암수소충전소    1  서울특별시   마포구     서울특별시 마포구 하늘공원로 108    
1        마곡에코수소충전소    1  서울특별시   강서구       서울특별시 강서구 양천로 225    
2          오곡수소충전소    1  서울특별시   강서구       서울특별시 강서구 벌말로 271    
3         H국회수소충전소    1  서울특별시  영등포구      서울특별시 영등포구 의사당대로 1    
4          양재수소충전소    1  서울특별시   서초구    서울특별시 서초구 바우뫼로12길 73    
..             ...  ...    ...   ...                      ...   
103  하이넷 통영용남수소충전소    1   경상남도   통영시   경상남도 통영시 용남면 남해안대로 38    
104        김해수소충전소    2   경상남도   김해시  경상남도 김해시 김해대로2596번길 53    
105  하이넷 거제장평수소충전소    1   경상남도   거제시      경상남도 거제시 거제대로 5050    
106     양산시증산수소충전소    1   경상남도   양산시    경상남도 양산시 물금읍 메기로 114    
107       H함안수소충전소    1   경상남도   함안군    경상남도 함안군 군북면 현포로 205    

                            이용가능시간  
0                      8~20시(연중무휴)  
1                    8~20시(일요일 휴무)  
2                      8~22시(연중무휴)  
3         평일 8~24시 / 주말, 공휴일 8~22시  
4                      8~22시(연중무휴)  
..                             ...  
103  7~21시(매주 수요일 / 둘, 넷째주 일요일 휴무)  
104       평일 8~21시 / 주말, 공휴일 9~18시  
105                  6~24시(일요일 휴무)  
106                    7~23시(연중무휴)  
107                    8~22시(연중무휴)  

[108 rows x 6 columns]

In [206]:
df_HVCS_group = df_HVCS.groupby('시').count()
df_HVCS_group

충전소  충전기   구  주소  이용가능시간
시                                
강원도        7    7   7   7       7
경기도       19   19  19  19      19
경상남도      12   12  12  12      12
경상북도       2    2   2   2       2
광주광역시      6    6   6   6       6
대구광역시      4    4   4   4       4
대전광역시      6    6   6   6       6
부산광역시      3    3   3   3       3
서울특별시      6    6   6   6       6
세종특별자치시    2    2   2   2       2
울산광역시      9    9   9   9       9
인천광역시      5    5   5   5       5
전라남도       3    3   3   3       3
전라북도       5    5   5   5       5
충청남도       9    9   9   9       9
충청북도      10   10  10  10      10

In [207]:
df_busan_HVCS = df_HVCS[df_HVCS['시']=='부산광역시'].reset_index(drop=True)
df_busan_HVCS

충전소  충전기      시    구                       주소  \
0  서부산NK수소충전소    1  부산광역시  강서구      부산광역시 강서구 가락대로 347    
1    H부산수소충전소    3  부산광역시  사상구      부산광역시 사상구 학감대로 167    
2  하이넷부산정관수소충    1  부산광역시  기장군  부산광역시 기장군 정관읍 신정관로 187    

                       이용가능시간  
0            8~18시(토, 일요일 휴무)  
1  버스 5~22시 / 넥쏘 10~22시(연중무휴)  
2               8~20시(월요일 휴무)

#### 수소차 충전소 수 적합도

In [249]:
HVCS_pro,HVCS_pro_negative = cal_norm(df_HVCS_group['충전소'].mean(),df_HVCS_group['충전소'].std(),
df_HVCS_group['충전소'].min(),df_HVCS_group['충전소'].max(),HVCS_standard,True)
HVCS_pro

0.5047886378261499

### 교통 편의성 데이터

In [443]:
# 주차장 정보
parking_intersection_file_path = 'C:\\Users\\SAMSUNG\\Desktop\\Green-Car-Charging-Station-Site-Selection\\Data\\부산 주차장 현황(교차로).csv'
df_parking_intersection = pd.read_csv(parking_file_path, encoding='cp949')
df_parking_intersection

Unnamed: 0      운영형태       관리기관명          주차장명     주차장관리번호  \
0             0    구(군)직영  부산광역시 사상구청         감전동주변  2010000013   
1             2      민간위탁  부산광역시 사상구청   엄궁초등학교지하주차장  2010000016   
2             3    구(군)직영  부산광역시 사상구청        일신주택주변  2011000003   
3             4    구(군)직영  부산광역시 사상구청       주례중학교주변  2012000004   
4             5      민간위탁  부산광역시 사상구청  모라1동행정복지센터주변  2007001725   
..          ...       ...         ...           ...         ...   
468         608      민간위탁   부산광역시 중구청   중앙동 노외공영주차장  2009000002   
469         610      민간운영  부산광역시 기장군청        1구역 2소  2008000002   
470         611      민간운영  부산광역시 기장군청        1구역 3소  2008000003   
471         612      민간운영  부산광역시 기장군청        3구역 1소  2008000006   
472         613  시설관리공단위탁  부산광역시 기장군청    기장시장 공영주차장  2018000001   

             소재지도로명주소 주차장구분 주차장유형  주차구획수         위도          경도  교차로  
0       사상구 감전동 589-2    공영    노외     26  35.154108  128.980071  239  
1    사상구 엄궁로191번길 23     공영    노외     93  35.128446  128.973403   85  
2      사상구 주례3동 545-3    공영    노외     10  35.145207  129.000672  165  
3      사상구 주례2동 82-13    공영    노외      8  35.154384  129.011131  276  
4      사상구 백양대로 949-6    공영    노상    156  35.171044  128.988085  249  
..                ...   ...   ...    ...        ...         ...  ...  
468   중구 대청로135번길 36     공영    노외     29  35.104229  129.034810  593  
469     기장군 기장읍 280-2    공영    노상     12  35.222141  129.206947   19  
470     기장군 기장읍 444-1    공영    노상     13  35.222141  129.206947   19  
471    기장군 기장읍 418-12    공영    노상     24  35.222141  129.206947   19  
472    기장군 읍내로94번길 9     공영    노외    216  35.244174  129.214471  341  

[473 rows x 12 columns]

#### 부산시 교차로수 적합도 구하기

In [445]:
df_parking_intersection['교차로'].mean()

248.4016913319239

In [449]:
intersection_pro, intersection_pro_negative = cal_norm(df_parking_intersection['교차로'].mean(),df_parking_intersection['교차로'].std(),
df_parking_intersection['교차로'].min(),df_parking_intersection['교차로'].max(),intersection_standard,False)

### 주변 충전소와의 거리 데이터

### 고속도로 여부

## 상위 요소 - 경제적 요소

### 전기차 충전기 정보 불러오기

### 충전소 설치 비용 불러오기 (전기차)
충전기 업체 마다 비용이 상이함
충전기 종류도 매우 많고 데이터 또한 없음

충전 사업 정부지원
- 1기 200만원, 2~5기 180만원, 6기 이상 150만원

총비용 
- 전기차 충전기 제품 비용 + 한전시설부담금

1. 파워큐브 - 비공용: 이동형 610,000원, 고정형 1,300,000원 / 
            - 공용: 주차면100면 이상 0원
2. kt - 충전기 가격 X
3. 지엔텔 - 충전기 가격 X
4. 에버온 - 충전기 가격 X
5. 삼성 - 비공용: 550,000원, 
        - 공용: (완속) 1,000,000원/1,600,000(2채널)  /  (급속) 58,000,000원(100kw), 40,000,000원(200kw)  
               

### 충전소 설치 비용 데이터 불러오기 (수소차) 
(250kg급 기준)

 파이프라인 방식 24억원

 튜브트레일러방식 22억원
 
 이동식 18억원
 
 천연가스개질방식은 46억원
 
 수전해방식은 31억원

천연가스 개질방식이 가장 비싸고, 이동식이 가장 저렴

In [214]:
HVCS_construction_cost_file_path = 'C:\\Users\\SAMSUNG\\Desktop\\Green-Car-Charging-Station-Site-Selection\\Data\\수소 충전소 종류별 구축 비용.csv'
df_HVCS_construction_cost = pd.read_csv(HVCS_construction_cost_file_path, encoding='cp949')
df_HVCS_construction_cost = df_HVCS_construction_cost.set_index('충전소 종류')
df_HVCS_construction_cost


구축 비용
충전소 종류              
파이프라인     2400000000
튜브트레일러방식  2200000000
이동식       1800000000
천연가스개질방식  4600000000
수전해방식     3100000000

#### 수소 충전소 구축 비용 적합도

In [257]:
HVCS_construction_cost_pro,HVCS_construction_cost_pro_negative  = cal_norm(df_HVCS_construction_cost['구축 비용'].mean(),df_HVCS_construction_cost['구축 비용'].std(),
df_HVCS_construction_cost['구축 비용'].min(),df_HVCS_construction_cost['구축 비용'].max(),HVCS_construction_cost_standard,True)
HVCS_construction_cost_pro

0.5379381229554602

### 총 유지비용

### LPG 충전소 토지비용
- LPG 충전소의 면적과 공시지가를 통해 토지비용 도출

In [216]:
cost = []
for i in range(df_LPG.shape[0]):
    area = df_LPG.iloc[i]['면적']
    price = df_LPG.iloc[i]['공시지가']
    cost.append(area * price)

df_LPG.insert(8,'토지비용',cost)
df_LPG

행정구역                      업소명                      소재지  \
0    부산 강서구               개인택시신공항충전소      부산 강서구 경전철로188번길 95   
1    부산 강서구               부산공항LPG충전소           부산 강서구 공항로 459   
2    부산 강서구                   신항만에너지         부산 강서구 낙동남로 507    
3    부산 강서구               르노삼성자동차(주)         부산 강서구 르노삼성대로 61   
4    부산 강서구                 죽림LPG충전소          부산 강서구 호계로 102    
5    부산 강서구                    가락충전소          부산 강서구 호계로 195    
6    부산 금정구                    (주)해연       부산 금정구 개좌로 225-15    
7    부산 금정구                  (주)신진가스          부산 금정구 반송로 433    
8    부산 기장군            마라톤가스(주)청강충전소   부산 기장군 기장읍 기장대로 343 A동   
9    부산 기장군               한국에너지산업(주)  부산 기장군 기장읍 기장대로413번길 38   
10   부산 기장군                  선암가스충전소    부산 기장군 장안읍 기장대로 1451    
11   부산 기장군    KH에너지(주)직영장안충전소(부산방향)     부산 기장군 장안읍 동해고속도로 25   
12   부산 기장군    KH에너지(주)직영장안충전소(울산방향)     부산 기장군 장안읍 동해고속도로 26   
13   부산 기장군                    정림에너지     부산 기장군 정관읍 신정관로 187    
14   부산 기장군        광신개발(주)직영 정관가스충전소       부산 기장군 정관읍 정관로 907   
15   부산 기장군               개인택시동부산충전소      부산 기장군 철마면 반송로 1022   
16    부산 동구             개인택시남부LPG충전소             부산 동구 성남로 42   
17    부산 동구  (주)대양씨앤씨(구:녹색에너지 범일영업소)           부산 동구 성남로 74-3   
18    부산 북구              개인택시 금곡역충전소           부산 북구 금곡대로 704   
19    부산 북구                    (주)삼보         부산 북구 만덕대로 327-0   
20   부산 사상구                  팔성가스충전소           부산 사상구 가야대로 75   
21   부산 사상구            (주)성은에너지제2충전소          부산 사상구 강변대로 412   
22   부산 사상구             극동유화(주)감전충전소          부산 사상구 낙동대로 878   
23   부산 사하구             광진기업(주)장림영업소           부산 사하구 다대로 309   
24   부산 사하구          광신가스(주)을숙도대로충전소         부산 사하구 을숙도대로 530   
25   부산 사하구             개인택시사하엘피지충전소         부산 사하구 을숙도대로 583   
26   부산 연제구             부산개인택시엘피지충전소         부산 연제구 월드컵대로 399   
27   부산 영도구                 (주)태종에너지           부산 영도구 영선대로 41   
28   부산 영도구                    부산에너지           부산 영도구 해양로 160   
29   부산 영도구            개인택시 영도LPG충전소           부산 영도구 해양로 296   
30  부산 해운대구             이원 해운대LPG충전소        부산 해운대구 해운대로 1255   

                관리구분         위도          경도       면적       공시지가          토지비용  
0              자동차충전  35.198669  128.966594   2545.0  2144000.0  5.456480e+09  
1              자동차충전  35.149064  128.951891   3105.0  1174000.0  3.645270e+09  
2              자동차충전  35.112947  128.875260   3105.0  1174000.0  3.645270e+09  
3              자동차충전  35.095408  128.883215   1538.0   640100.0  9.844738e+08  
4              자동차충전  35.200736  128.893462   1580.0  1017000.0  1.606860e+09  
5              자동차충전  35.208847  128.890148   2600.0  1017000.0  2.644200e+09  
6       용기+탱크+13kg용기  35.230796  129.122388   3346.0  1370000.0  4.584020e+09  
7              자동차충전  35.215006  129.111143   1646.4  2797000.0  4.604981e+09  
8              자동차충전  35.225657  129.213988   2592.0  1069000.0  2.770848e+09  
9   용기+자동차+탱크+13kg용기  35.235543  129.215629   2473.0  1063000.0  2.628799e+09  
10        자동차+13kg용기  35.321585  129.241320   2717.0  1060000.0  2.880020e+09  
11             자동차충전  35.377525  129.246402  15637.0   263500.0  4.120350e+09  
12             자동차충전  35.382870  129.248394  37294.0   263500.0  9.826969e+09  
13             자동차충전  35.328714  129.201710   2575.0   743500.0  1.914512e+09  
14        자동차+13kg용기  35.324749  129.210964   1899.0   877400.0  1.666183e+09  
15             자동차충전  35.235930  129.163728   2543.0   923000.0  2.347189e+09  
16             자동차충전  35.130627  129.054917   1518.5  1701000.0  2.582968e+09  
17         용기+13kg용기  35.130891  129.057587   3169.8  1545000.0  4.897341e+09  
18             자동차충전  35.269475  129.018781   2055.0   842500.0  1.731338e+09  
19             자동차충전  35.213643  129.040635   3837.0  1608000.0  6.169896e+09  
20        자동차+13kg용기  35.151317  128.980986   2946.6  3247000.0  9.567610e+09  
21             자동차충전  35.122092  128.962063   3732.4  2373000.0  8.856985e+09  
22  용기+자동차+탱크+13kg용기  35.138888  128.971339   2886.1  3201000.0  9.238406e+09  
23    

#### LPG 충전소 토지비용 적합도

In [274]:
LPG_cost_pro,LPG_cost_pro_negative = cal_norm(df_LPG['토지비용'].mean(),df_LPG['토지비용'].std(),
df_LPG['토지비용'].min(),df_LPG['토지비용'].max(),LPG_cost_standard,True)
LPG_cost_pro

0.7528610639760787

### 주차 구획 수 데이터 불러오기

In [378]:
parking_file_path = 'C:\\Users\\SAMSUNG\\Desktop\\Green-Car-Charging-Station-Site-Selection\\Data\\부산 주차장 현황.csv'
df_parking = pd.read_csv(parking_file_path, encoding='cp949')
df_parking

운영형태       관리기관명            주차장명     주차장관리번호          소재지도로명주소 주차장구분  \
0      구(군)직영  부산광역시 사상구청           감전동주변  2010000013     사상구 감전동 589-2    공영   
1      구(군)직영  부산광역시 사상구청           장성주차장  2010000015   사상구 주례2동 92-278    공영   
2        민간위탁  부산광역시 사상구청     엄궁초등학교지하주차장  2010000016  사상구 엄궁로191번길 23     공영   
3      구(군)직영  부산광역시 사상구청          일신주택주변  2011000003    사상구 주례3동 545-3    공영   
4      구(군)직영  부산광역시 사상구청         주례중학교주변  2012000004    사상구 주례2동 82-13    공영   
..        ...         ...             ...         ...               ...   ...   
609    구(군)직영  부산광역시 기장군청  정관읍 상업지역 노상주차장  2016000017       기장군 정관면 732    공영   
610      민간운영  부산광역시 기장군청          1구역 2소  2008000002     기장군 기장읍 280-2    공영   
611      민간운영  부산광역시 기장군청          1구역 3소  2008000003     기장군 기장읍 444-1    공영   
612      민간운영  부산광역시 기장군청          3구역 1소  2008000006    기장군 기장읍 418-12    공영   
613  시설관리공단위탁  부산광역시 기장군청      기장시장 공영주차장  2018000001    기장군 읍내로94번길 9     공영   

    주차장유형  주차구획수         위도          경도  
0      노외     26  35.154108  128.980071  
1      노외     16   0.000000    0.000000  
2      노외     93  35.128446  128.973403  
3      노외     10  35.145207  129.000672  
4      노외      8  35.154384  129.011131  
..    ...    ...        ...         ...  
609    노상     46   0.000000    0.000000  
610    노상     12  35.222141  129.206947  
611    노상     13  35.222141  129.206947  
612    노상     24  35.222141  129.206947  
613    노외    216  35.244174  129.214471  

[614 rows x 10 columns]

### 주차 구획수 통계 구하기

In [219]:
df_parking_area = df_parking['주차구획수'].astype(int)
df_parking_area.describe()

count    615.000000
mean      62.378862
std       70.372193
min       -4.000000
25%       18.000000
50%       38.000000
75%       79.000000
max      414.000000
Name: 주차구획수, dtype: float64

#### 주차구획 수 적합도

In [276]:
parking_area_pro,parking_area_pro_negative = cal_norm(df_parking_area.mean(),df_parking_area.std(),df_parking_area.min(),df_parking_area.max(),parking_area_standard,False)
parking_area_pro

0.32246243546520803

## 상위 요소 - 기술적 요소

### 복합 충전소 여부 (수소차)

기존의 LPG 충전소 부지에 수소충전소를 세우는 경우 "적합" 100%로 설정, 반대의 경우 "부적합" 100%로 설정함

### 충전시간에 따른 충전방식 (전기차, 정성적) 
 유동인구가 해당지역의 거주인구보다 많은 지역에 해당하는 경우 급속충전의 이용율이 높음
 
 따라서 "급속충전" 100%로 설정, 반대의 경우는 "완속충전" 100%로 설정함  

In [221]:
# 유동인구와 고정인구에 따른 충전방식 적합도 계산 
def charge_type(foottraffic_pro,busan_people_pro):
    '''
    foottraffic_pro : 유동인구 적합도
    busan_people_pro : 고정인구 적합도
    charger_pro : 설정하고자 하는 충전기 타입의 적합도 비율 초기값(0~1)
    '''
    # 초기 급촉 충전기 타입의 확률는 0.5  
    charger_pro = 0.38

    if foottraffic_pro >= busan_people_pro:
        charger_type = 'Fast Charge'
    else:
        charger_type = 'Standard Charge'

    # 유동인구와 기존인구와 차이가 나는 만큼 더 많은 확률 부여
    if foottraffic_pro >= busan_people_pro:
        charger_type = 'Fast Charge'
        diff = foottraffic_pro - busan_people_pro
        charger_pro += diff
        if charger_pro > 1:
            charger_pro = 1.0
        print('Fast Charge:', charger_pro)
        print('Standard Charge:', 1-charger_pro)
    elif foottraffic_pro < busan_people_pro:
        charger_type = 'Standard Charge'
        diff = busan_people_pro - foottraffic_pro
        charger_pro -= diff
        if charger_pro < 0:
            charger_pro = 0.0
        print('Fast Charge:', charger_pro)
        print('Standard Charge:', 1-charger_pro)
    return charger_type, charger_pro, 1-charger_pro

#### 사하구의 전기차 충전소의 타입

In [222]:
charger_type,charger_pro_fast,charger_pro_standard = charge_type(foottraffic_pro,busan_people_pro)
print(charger_type)

Fast Charge: 0.36410058719061145
Standard Charge: 0.6358994128093886
Standard Charge


## 결합 확률로 상위 요인 도출하기


In [450]:
# E는 전기차, H는 수소차
variable_environment = [] # 환경변수는 전기/수소 동일
variable_social_E = []
variable_social_H = []
variable_economic_E = []
variable_economic_H = []
variable_technical_E = []
variable_technical_H = []

# 가중치
weight_environment = [0.13,0.13,0.13,0.13,0.13,0.13,0.22] # co, so2, no2, ozone, pm10, pm25, 통합대기환경지수(khai)
weight_social_E = [0.1,0.1,0.3,0.3,0.2] # 고정인구, 유동인구, 전기차 수, 전기차 충전소 수, 교통 편의성
weight_social_H = [0.1,0.1,0.1,0.3,0.2,0.2] # 고정인구, 유동인구, LPG 충전소 수, 수소차 수, 수소차 충전소 수, 교통 편의성 
weight_economic_E = [1.0] # 주차 구획수
weight_economic_H = [0.5,0.5] # LPG 충전소 토지비용, 수소충전소 구축 비용
weight_technical_E = [1.0] # 급속/완속 적합성
weight_technical_H = [] # 
weight_primary_factor = [0.15, 0.4, 0.3, 0.15] # 상위 요소들의 가중치(환경, 사회, 경제, 기술)

variable_environment.append(co_pro)
variable_environment.append(so2_pro)
variable_environment.append(no2_pro)
variable_environment.append(o3_pro)
variable_environment.append(pm10_pro)
variable_environment.append(pm25_pro)
variable_environment.append(khai_pro)


variable_social_E.append(busan_people_pro)
variable_social_E.append(foottraffic_pro)
variable_social_E.append(elec_vehicle_pro)
variable_social_E.append(EVCS_pro)
variable_social_E.append(intersection_pro)

variable_social_H.append(busan_people_pro)
variable_social_H.append(foottraffic_pro)
variable_social_H.append(LPG_pro)
variable_social_H.append(hidro_vehicle_pro)
variable_social_H.append(HVCS_pro)
variable_social_H.append(intersection_pro)


variable_economic_E.append(parking_area_pro)

variable_economic_H.append(LPG_cost_pro)
variable_economic_H.append(HVCS_construction_cost_pro)


variable_technical_E.append(charger_pro_fast)


print("환경변수(전기,수소)",variable_environment)
print("사회변수(전기)",variable_social_E)
print("사회변수(수소)",variable_social_H)
print("경제변수(전기)",variable_economic_E)
print("경제변수(수소)",variable_economic_H)
print("기술변수(전기)",variable_technical_E)
print("기술변수(수소)",variable_technical_H)

환경변수(전기,수소) [0.8287531282163694, 0.7911203783441816, 0.909590433996196, 0.8017202428389887, 0.8866811666335976, 0.8415688322677302, 0.8220286272881268]
사회변수(전기) [0.5383756609850885, 0.5848617448890694, 0.6626879332080854, 0.6344991485975281, 0.8271541215787344]
사회변수(수소) [0.5383756609850885, 0.5848617448890694, 0.7148108415065166, 0.4304353506049742, 0.5047886378261499, 0.8271541215787344]
경제변수(전기) [0.32246243546520803]
경제변수(수소) [0.7528610639760787, 0.5379381229554602]
기술변수(전기) [0.36410058719061145]
기술변수(수소) []


### 결합확률 함수

In [451]:
# 결합 확률 구하기
def combine_pro(list, weight):
    result = 0.0
    for pro, wei in list, weight:
        result += pro*wei
    return result

### 조건부확률분포표

 - 환경적 요인

    오존, 아황산가스, 이산화질소, 일산화탄소, 미세먼지(pm10, pm25)

 - 사회적 요인
   
    고정인구수, 유동인구 수, 친환경 자동차 등록대수, LPG 충전소 수, 충전소 수(전기,수소), 교통 편의성, 주변 충전소와의 거리, 고속도로 여부

 - 경제적 요인
   
    전기차 충전기 설치 비용, 수소 충전소 구축 비용, LPG 충전소 토지비용, 총 유지비용, 주차 구획 수

 - 기술적 요소

    복합 충전소 여부(수소), 충전시간에 따른 급속/완속 충전 방식

In [439]:
# 조건부 확률 구하기
def conditional_pro(weight):
    '''
    weight : 결합할 변수의 가중치 값(list)
    '''
    table_len = 2**len(weight)
    result_True = [0.0]*table_len
    for i in range(len(weight)): # 3번
        for j in range(table_len): # 8번
            cycle = table_len/(2**(i+1))
            if int(j/cycle)%2 == 0:
                result_True[j] += weight[i]
                # 부동소수점 오차 제거
                result_True[j] = round(result_True[j],6)

    result_False = list(reversed(result_True))

    return result_True, result_False


In [441]:
# 상위 요인 조건부 확률
environment_pro, environment_pro_negative = conditional_pro(weight_environment)
social_E_pro, social_E_pro_negative = conditional_pro(weight_social_E)
social_H_pro, social_H_pro_negative = conditional_pro(weight_social_H)
economic_E_pro, economic_E_pro_negative = conditional_pro(weight_economic_E)
economic_H_pro, economic_H_pro_negative = conditional_pro(weight_economic_H)
technical_E_pro, technical_E_pro_negative = conditional_pro(weight_technical_E)

print("환경적 요소 조건부 확률",'\n', environment_pro,'\n',environment_pro_negative)
print("사회적(전기) 요소 조건부 확률",'\n', social_E_pro,'\n',social_E_pro_negative)
print("사회적(수소) 요소 조건부 확률",'\n', social_H_pro,'\n',social_H_pro_negative)
print("경제적(전기) 요소 조건부 확률",'\n', economic_E_pro,'\n',economic_E_pro_negative)
print("경제적(수소) 요소 조건부 확률",'\n', economic_H_pro,'\n',economic_H_pro_negative)
print("기술적(전기) 요소 조건부 확률",'\n', technical_E_pro,'\n',technical_E_pro_negative)

환경적 요소 조건부 확률 
 [1.0, 0.78, 0.87, 0.65, 0.87, 0.65, 0.74, 0.52, 0.87, 0.65, 0.74, 0.52, 0.74, 0.52, 0.61, 0.39, 0.87, 0.65, 0.74, 0.52, 0.74, 0.52, 0.61, 0.39, 0.74, 0.52, 0.61, 0.39, 0.61, 0.39, 0.48, 0.26, 0.87, 0.65, 0.74, 0.52, 0.74, 0.52, 0.61, 0.39, 0.74, 0.52, 0.61, 0.39, 0.61, 0.39, 0.48, 0.26, 0.74, 0.52, 0.61, 0.39, 0.61, 0.39, 0.48, 0.26, 0.61, 0.39, 0.48, 0.26, 0.48, 0.26, 0.35, 0.13, 0.87, 0.65, 0.74, 0.52, 0.74, 0.52, 0.61, 0.39, 0.74, 0.52, 0.61, 0.39, 0.61, 0.39, 0.48, 0.26, 0.74, 0.52, 0.61, 0.39, 0.61, 0.39, 0.48, 0.26, 0.61, 0.39, 0.48, 0.26, 0.48, 0.26, 0.35, 0.13, 0.74, 0.52, 0.61, 0.39, 0.61, 0.39, 0.48, 0.26, 0.61, 0.39, 0.48, 0.26, 0.48, 0.26, 0.35, 0.13, 0.61, 0.39, 0.48, 0.26, 0.48, 0.26, 0.35, 0.13, 0.48, 0.26, 0.35, 0.13, 0.35, 0.13, 0.22, 0.0] 
 [0.0, 0.22, 0.13, 0.35, 0.13, 0.35, 0.26, 0.48, 0.13, 0.35, 0.26, 0.48, 0.26, 0.48, 0.39, 0.61, 0.13, 0.35, 0.26, 0.48, 0.26, 0.48, 0.39, 0.61, 0.26, 0.48, 0.39, 0.61, 0.39, 0.61, 0.52, 0.74, 0.13, 0.35, 0.26, 0.48,